Replace dictionary in Primitives with tuple and turn Primitives into frozen dataclass

jakobj opened the pull request:
Instead of using an `OrderedDict` like suggested in #107 this PR replaces the dictionary used in `Primitives` with a list. Since function keys are just integers, they might as well be indices in an array. This automatically avoid a potential relabeling of functions when a `Primitives` object is converted to a list a reinitialized.

Also renames `Primitives.sample` -> `Primitives.sample_allele`.
Also removes some outdated comments.

closes #107 
closes #106

mschmidt87 commented on gp/primitives.py:
```diff
@@ -26,33 +23,24 @@ def __init__(self, primitives: List[Type[Node]]) -> None:
-        # unfortunately not supported by pickle, necessary for
-        # multiprocessing; another way to implement this?
-        # self._primitives = types.MappingProxyType(self._primitives)
+        self._primitives = list(primitives)
```
If `primitives` is already a list, why calling `list()` here?

jakobj replied:
to create a shallow copy of a list

mschmidt87 replied:
Got it.

mschmidt87 commented on gp/primitives.py:
```diff
@@ -26,33 +23,24 @@ def __init__(self, primitives: List[Type[Node]]) -> None:
 
-        self._determine_max_arity()
+    def __iter__(self) -> Iterator[Type[Node]]:
+        return iter(p for p in self._primitives)
```
why not `iter(self._primitives)`?

jakobj replied:
good point. but even better: we don't need this function any more :)

mschmidt87 replied:
Hmm, I think we do need it so that `list(Primitives)` works. That's also why `mypy` complains in CI.

jakobj replied:
it's not needed for `list()` (`__getitem__`) is enough, but i will look into the mypy issue

mschmidt87 commented on gp/primitives.py:
```diff
@@ -26,33 +23,24 @@ def __init__(self, primitives: List[Type[Node]]) -> None:
-        # multiprocessing; another way to implement this?
-        # self._primitives = types.MappingProxyType(self._primitives)
+        self._primitives = list(primitives)
+        self._max_arity: int = self._determine_max_arity()
```
Wouldn't this be better written as a class property then?

jakobj replied:
what is "this"? and why?

mschmidt87 replied:
`_max_arity`. This seems to me like the classic example where you want to use a class property: When accessing `_max_arity`, you want to call a routine that determines it on the fly. Or am I missing something?

jakobj replied:
once initialized, this class is/should be treated as constant, the `max_arity` hence does not change afterwards. to save runtime it hence makes sense to compute it once during initialization

mschmidt87 commented on gp/primitives.py:
```diff
@@ -62,20 +50,18 @@ def sample(self, rng: np.random.RandomState) -> int:
 
     def __getitem__(self, key: int) -> Type[Node]:
+        if key < 0 or key >= len(self._primitives):
+            raise IndexError("primitives index out of bounds")
```
"primitive index out of bounds"

jakobj replied:
yes!

jakobj commented:
thanks for the review :sparkles:  addressed all your comments, pls have another look

mschmidt87 commented:
I thought about the problem that an instance of `Primitives` is still mutable, potentially causing a risk in user applications. Here is my suggestion to solve that problem:
1. Instead of using a `list`, use a `tuple` for the `self._primitives` member, because `tuple` is immutable.
2. Turn the `Primitives` class into a `dataclass` and make it frozen, like so:
```python
@dataclass(frozen=True)
class Primitives:
    primitives: tuple(Type[Node])
```
You will need to adapt the class code a little bit because the `__init__` is then created automatically, i.e., everything that is not in `__init__` has to go to a `__post_init__` function. But that should be all and pretty straightforward.

What this will do is it will prevent two ways to manipulate the `_primitives` after instance creation. You can no longer do
```python
prim = Primitives((gp.Add, gp.Sub))
prim._primitives[0] = gp.Mul
```
because tuples are immutable.
And you can no longer do
```python
prim = Primitives((gp.Add, gp.Sub))
prim._primitives = (gp.Mul, gp.Pow)
```
because the members of `Primitive` are frozen.

jakobj commented:
ok, i've implemented your suggestion for a frozen dataclass. replacing the list with a tuple requires an API change, which subsequently required many changes in the tests. i've also added the additional dependency on "dataclasses" for python 3.6. please have another look @mschmidt87
